Update specification to make API more friendly

wusyong opened the pull request:
Address several changes to make it more user friendly, so they don't need to know too many details when using high level APIs:

- All `BIP32Path` parameters are optional. Since address derivation rfc doesn't come out yet. I would like to remain as is until the certain method is defined.
- `indexation` parameters is splited into `indexation_key` and `data`.
- Move GET method of address, output, and message back to low level. Users probable just want to use find_* API most of the time.

The amount must still be non zero is because transaction payload must have an output and its amount must not be zero during [syntactical validation](https://github.com/luca-moser/protocol-rfcs/blob/signed-tx-payload/text/0000-transaction-payload/0000-transaction-payload.md#syntactical-validation).
And the last thing I haven't updated yet is `reattach()`. @fijter  Do you mean it should be moved to `send()` method as optional parameter?

Commit:
Make BIP32Path optional in spec

Commit:
Update indexation parameter of send method in spec

Commit:
Move get message and output to low level

Commit:
Move get address to low level

Commit:
Rename get_addresses to find_addresses

fijter commented on specs/iota-rs-ENGINEERING-SPEC-0000.md:
```diff
@@ -156,7 +157,7 @@ Following are the steps for implementing this method:
 * Return the address with corresponding index on the wallet chain;
 
-## `get_addresses()`
+## `find_addresses()`
```
This function can probably be lower level instead of higher level - On a higher level you probably don't care about this, you just want to send and receive coins/data.

fijter commented on specs/iota-rs-ENGINEERING-SPEC-0000.md:
```diff
@@ -79,11 +79,12 @@ A generic send function for easily sending a value transaction message.
 | **output** | ✘ | Output | Users can manually pick their own output instead of having node decide on which output should be used. |
-| **indexation** | ✘ | Indexation | An optional indexation payload with indexation key and data. Both fields can be optional too. |
+| **indexation_key** | ✘ | String | An optional indexation key of the indexation payload. |
+| **data** | ✘ | String | An optional indexation data of the indexation payload. |
```
Semantics but I think both can be bytes (a String is just UTF-8 encoded bytes ofc, data or key can be in a binary representation as well iirc.)

wusyong replied:
Their type definition in RFC are string and bytes array respectively. I would change the data parameter to u8 bytes.

fijter commented on specs/iota-rs-ENGINEERING-SPEC-0000.md:
```diff
@@ -253,63 +254,6 @@ Following are the steps for implementing this method:
 * Perform proof-of-work;
 * Store messages on the tangle using [`post_messages()`](#post_messages-post-messages);
 
-## `get_message()`
```
Should probably stay in the higher level since this is a common thing to do, retrieving a single message by message id.

fijter reviewed (commented):
In addition to the comments in the code some other remarks:

 - Network in the constructor is not required, but you need to either have a network or a node/nodes/node_pool_urls list, one of those is mandatory but it doesn't matter which one.
 - `send` without value should be possible, in that case you don't send a Message with SignedTransaction with Indexation Payload but a Message with just a Indexation Payload instead and the Data in the value field of the indexation.
 - Maybe name message_id `message_ids` in `find_messages` since we always expect a list of MessageId's there?
 - Reattach could be a function on the message object returned by get_message/find_messages/send instead of a seperate api function. Does not make sense to add this as a optional attribute to send, you want to be able to re-attach any message regardless of if you've sent it yourself and if you actually called the send function in code or not.
 - find_ouputs and maybe find_addresses can probably move to lower level libs since I don't see this as something someone would commonly do on the higher level.

Commit:
Make network param optional

Commit:
Address review comments

- Move find_* method to low level
- Move get_message to high level
- Change params in send method to be optional
- Update type of data in send to u8 bytes

fijter reviewed (commented):
Almost there I guess, some final points:

 - `find_messages` back to Higher level API part please, that was okay there (just leave find_outputs in Full Node API part)
 - `get_message` should only be lookup by message_id, not index (since one index can have multiple messages, find_messages does allow index lookup, corrosponding to the API docs for this at /api/v1/messages/{messageId})
 - The type of `get_address_balances` `addresses` parameter is `[MessageId]`, should probably be [Address]

Commit:
Address more review cooments

- Move find_messages back to high level
- get_message now only takes message_id as parameter.
- address parameter in get_address_balances is the type of Address.

fijter reviewed (approved):
Nice, looks good to me at this point. Looking forward to any feedback from others utilizing this to see if we missed anything further.

fijter commented:
Ah one more thing I just noticed; `output` should be `outputs` for the `send` method since you can provide an array of multiple outputs to use, not just a single one.

cvarley100 committed:
Update iota-rs-ENGINEERING-SPEC-0000.md

cvarley100 committed:
Update iota-rs-ENGINEERING-SPEC-0000.md

cvarley100 commented:
@fijter I have updated the `output` and `address` parameters to arrays to allow for multiple. I suggest we keep the name the same.

lucasfernog committed:
Merge branch 'dev' into spec

lucasfernog commented:
When can we get the `reattach` function back?

lucasfernog commented:
Also we need to do something about this `is_confirmed` function - the wallet library needs to know that but it's not part of the node API spec (?)

Thoralf-M commented:
Why is `get_message` in the General API list and not in the Full Node API list?
And `find_addresses` is in the Full Node API, but it doesn't require the API at all

Commit:
chore: update specification with retry, reattach & promote methods

laumair commented:
> When can we get the `reattach` function back?

@lucasfernog Please see https://github.com/iotaledger/iota.rs/issues/151

> Also we need to do something about this is_confirmed function - the wallet library needs to know that but it's not part of the node API spec (?)

I do not think there will be a separate `is_confirmed` function. I believe [get_message](https://github.com/iotaledger/iota.rs/blob/dev/specs/iota-rs-ENGINEERING-SPEC-0000.md#get_message) when passed with `metadata()` option should provide us with the confirmation details of a message. The expected response interface should be:

```
{
    /**
     * The message id.
     */
    messageId: string;
    /**
     * The message id of parent 1.
     */
    parent1MessageId: string;
    /**
     * The message id of parent 2.
     */
    parent2MessageId: string;
    /**
     * Is the message solid.
     */
    isSolid?: boolean;
    /**
     * Is the message referenced by a milestone.
     */
    referencedByMilestoneIndex?: number;
    /**
     * The ledger inclusion state.
     */
    ledgerInclusionState?: string;
    /**
     * Should the message be promoted.
     */
    shouldPromote?: boolean;
    /**
     * Should the message be reattached.
     */
    shouldReattach?: boolean;
}
```

laumair commented:
@fijter @cvarley100 I couldn't find any information about address checksums. I believe there should be support for creating checksums for addresses (maybe on the lower level).

lucasfernog committed:
feat(message) add missing metadata fields

lucasfernog committed:
chore(client) remove `is_confirmed` API

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into spec

Commit:
Merge remote-tracking branch 'upstream/spec' into chore/retry-method-spec

bingyanglin committed:
Merge pull request #150 from laumair/chore/retry-method-spec

Update specification with retry, reattach & promote methods

bingyanglin committed:
Add quorum operation session

bingyanglin commented on specs/iota-rs-ENGINEERING-SPEC-0000.md:
```diff
@@ -69,6 +69,34 @@ The data structure to initialize the instance of the Higher level client library
+
+* Unsolved Question:
+  - We need to define the list of APIs (e.g., `get_balance()`, `get_address_balances`, etc.), whose returned values should depend on the `quorum_size` nodes.
+
```
Listed the open questions

bingyanglin committed:
Correct parameter/returned types in spec

bingyanglin committed:
Merge pull request #160 from bingyanglin/spec

Correct parameter/returned types in spec

bingyanglin committed:
Enhance the return object of retry()/reattach()/promote() and add more description

bingyanglin committed:
Modify the retry() description

bingyanglin committed:
Merge pull request #168 from bingyanglin/spec

Enhance the return object of retry()/reattach()/promote() and the description

bingyanglin committed:
Fix typo

fijter committed:
Update iota-rs-ENGINEERING-SPEC-0000.md

Updated the spec with mqtt event wrapper, additional timeout features, default values in arguments, removed the quorum functionality for now, and other small improvements.

fijter committed:
Update iota-rs-ENGINEERING-SPEC-0000.md

Removed unresolved questions section

bingyanglin committed:
Merge pull request #169 from fijter/patch-2

Update iota-rs-ENGINEERING-SPEC-0000.md

fijter committed:
Fixed broken tables and consistency 

Some tables were broken in the rendering, those are fixed. I also changed the `node_sync_interval` to be in ms instead of seconds for consistency with the other time values.

bingyanglin committed:
Merge pull request #171 from fijter/patch-4

Fixed broken tables and consistency

lucasfernog committed:
Merge branch 'dev' into spec
